### PR TITLE
Create all file information

### PR DESCRIPTION
**Pull request by MancunianSam**
This was originally going to just be the checksum calculation but I needed most of the functions already defined for the second part and actually extracting the other fields is relatively simple so I've included them all in here. 

I've also added in a callback function so we can get the progress as they are processed. This is optional in case you're not interested in the progress. At the moment, it returns a very simple `{totalFiles: 1, processedFiles: 1` but if we need it to provide any more information later then we can add that in.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { bytes_to_hex, Sha256 } from "asmcrypto.js"
+import { TFileInformation, TdrFile } from "./types"
+
+export const getFileInformation: TFileInformation = async (
```
Comment (suzannehamilton) [inline src/index.ts]: I wonder if we should call it `fileMetadata` rather than `fileInformation`? David C might have an opinion on this, especially if we're releasing this a general-purpose library.

I also wonder if there's a better word than `get`, which sounds quite Java-y? `extract`? `calculate`?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  files,
```
Comment (suzannehamilton) [inline src/index.ts]: If this takes an array of files, presumably the client will have to do the work of recursively iterating through a directory? I wonder if this library should take care of that too?

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  progressFunction
+) => {
+  let processedFiles = 1
```
Comment (suzannehamilton) [inline src/index.ts]: I know it was me who said that it would be clearer to report on the number of files processed than the percentage, but now I'm wondering whether we should report on both?

Calculating checksums for big files can be quite slow (see the [results from Alpha](https://github.com/nationalarchives/tdr-dev-documentation/blob/master/technology-considerations/frontend_checksum_options.md#results)), so I think it would be helpful for the progress bar to keep moving as we process a big file.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  for (let i = 0; i < chunkCount; i += 1) {
+    const start = i * chunkSize
+    const end = start + chunkSize
+    const slice: Blob = file.slice(start, end)
+    sha256.process(await sliceToUintArray(slice))
+  }
+
+  const result = sha256.finish().result!
+  return bytes_to_hex(result)
```
Comment (suzannehamilton) [inline src/index.ts]: Probably worth adding `.trim()` like we [did in Alpha](https://github.com/nationalarchives/tdr-prototype-mvc/pull/81) to fix the Safari bug, just in case we can't fix it in the original library.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  return await new Promise(resolve => {
+    fileReader.onload = () => {
+      const { result } = fileReader
+      if (result instanceof ArrayBuffer) {
+        resolve(new Uint8Array(result))
+      }
+    }
+  })
+}
+export const generateHash: (file: File) => Promise<string> = async file => {
```
Comment (suzannehamilton) [inline src/index.ts]: At some point (probably not now) we should consider whether we should use the built-in subtle.crypto for small files. It was [quite a lot faster](https://github.com/nationalarchives/tdr-dev-documentation/blob/master/technology-considerations/frontend_checksum_options.md#results) when processing lots of small files.

File: src/index.ts (hunk 2 above)
Comment (MancunianSam) [reply src/index.ts]: We don't need to iterate over the directory if we stick with an file input element, you only need to do `event.currentTarget.fileinputid.files` and it lists everything. I think the iteration was for the drag and drop which hopefully we're not using again. 
If we leave it as an array of files, it will make splitting up the requests easier if we get a large number of files, we can pass in slices of the full arary individually It. makes it a bit more generic for others to use as well in that you can pass any array of File like objects and it will work (ish)

File: src/index.ts (hunk 1 above)
Comment (MancunianSam) [reply src/index.ts]: Yes, that's much better :smile: I was having real trouble naming things. I've called the function `extractFileMetadata` and  I've renamed the repo `tdr-file-metadata` as well

File: src/index.ts (hunk 3 above)
Comment (MancunianSam) [reply src/index.ts]: I've changed this now so that the percentage is updated each time a slice of a file is processed. This should give us a reasonably smooth progress bar.

File: src/index.ts (hunk 4 above)
Comment (MancunianSam) [reply src/index.ts]: Ok done. I might have a look at that library next learning day perhaps.

File: src/index.ts (hunk 5 above)
Comment (MancunianSam) [reply src/index.ts]: I agree, once we start testing it with lots of files, we can have a look at performance, as you say, I'll leave it for now.

File: src/index.ts (hunk 2 above)
Comment (suzannehamilton) [reply src/index.ts]: Ah, I'd forgotten the recursion was just in the drag-and-drop. Thanks!

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+) => {
+  let processedChunks = 0
+  const totalChunks = getTotalChunks(files, chunkSize)
+
+  const chunkProgress: TChunkProgressFunction = () => {
+    processedChunks += 1
+    const ratioProcessed = processedChunks / totalChunks
+    const percentageProcessed = Math.round(ratioProcessed * 100)
+    const totalFiles = files.length
+    const processedFiles = Math.floor(ratioProcessed * totalFiles)
```
Comment (suzannehamilton) [inline src/index.ts]: This number's a bit misleading because it's a kind of average. Can we keep track of the actual number of processed files?

This is why I was a bit apprehensive about making this a library from the start - we're sort of guessing about what we need. :slightly_smiling_face:

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  TGenerateMetadata,
+  TTotalChunks,
+  TSliceToArray,
+  TChunkProgressFunction
+} from "./types"
+
+export const extractFileMetadata: TFileMetadata = async (
+  files,
+  progressFunction,
+  chunkSize = 100000000
```
Comment (suzannehamilton) [inline src/index.ts]: Could this be `chunkSizeBytes`. It will probably save some confusion.

File: test/index.test.ts [hunk 8]
```diff
@@ -1 +1,71 @@
-test('returns the correct checksum for a file', async () => {});
+import { extractFileMetadata } from "../src/index"
+import * as fs from "fs"
+
+test("returns the correct checksum for a file", async () => {
+  const blob = new Blob([new Uint8Array(fs.readFileSync("test/testfile"))])
+
+  const result = await extractFileMetadata([new File([blob], "")])
+  expect(result[0].checksum).toEqual(
+    "e2d0fe1585a63ec6009c8016ff8dda8b17719a637405a4e23c0ff81339148249"
```
Comment (suzannehamilton) [inline test/index.test.ts]: It would be good to test to have a test that this checksum is independent of the chunk size, to catch issues with chunking.

File: .eslintrc.json [hunk 9]
```diff
@@ -1,30 +1,28 @@
...
+    "SharedArrayBuffer": "readonly"
+  },
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "ecmaVersion": 2018,
+    "sourceType": "module"
+  },
+  "plugins": ["@typescript-eslint", "jest", "prettier"],
+  "rules": {
+    "jest/no-disabled-tests": "warn",
```
Comment (suzannehamilton) [inline .eslintrc.json]: This reminds me - do we have a Jenkins build which runs eslint and tests?

File: .eslintrc.json [hunk 10]
```diff
@@ -1,30 +1,28 @@
...
-        "jest/no-focused-tests": "error",
-        "jest/no-identical-title": "error",
-        "jest/prefer-to-have-length": "warn",
-        "jest/valid-expect": "error"
-    }
-}
\ No newline at end of file
+  "env": {
+    "browser": true,
+    "es6": true,
```
Comment (suzannehamilton) [inline .eslintrc.json]: Silly JavaScript question: what does this setting mean if the project uses Typescript?

Comment (suzannehamilton) [reply .eslintrc.json]: Follow-up question: if we'd like to make this library generally available, are we building it in a way that can be used cross-browser? (e.g. without ES6 polyfills.)

File: src/index.ts (hunk 6 above)
Comment (MancunianSam) [reply src/index.ts]: It's not average, `Math.floor` makes it exactly right. It is quite a lot more difficult to pass around the number of processed files variables and as this produces exactly the same result,

File: src/index.ts (hunk 7 above)
Comment (MancunianSam) [reply src/index.ts]: Yeah, makes sense

File: test/index.test.ts (hunk 8 above)
Comment (MancunianSam) [reply test/index.test.ts]: Yeah, that's true, I'll add one in.

File: .eslintrc.json (hunk 9 above)
Comment (MancunianSam) [reply .eslintrc.json]: There's no Jenkinsfile at all yet, I was going to do it separately to this PR because I need something beyond the basic skeleton to actually publish to npm.

File: .eslintrc.json (hunk 10 above)
Comment (MancunianSam) [reply .eslintrc.json]: This is just the linter config but you have a point about older browsers, I've set the target to ES5 in tsconfig.json so this should put in the ES6 polyfils we need.

File: src/index.ts (hunk 6 above)
Comment (MancunianSam) [reply src/index.ts]: I've just had a think about this actually and I think you're right, let me check it.

Comment (MancunianSam) [reply src/index.ts]: I think the problem with this is that because getting the byte array from the file is async, the percentage progress, i.e. how many chunks have been processed and how many files have been processed, i.e. how often the generateHash function completes, are not always going to be in sync because the function can complete before every await inside the loop has finished processing. If I try passing the number of files variable in, it doesn't always match up with the percentage because of this. 
In short, I don't think we can have both and I think percentage is more useful because it will still increment regularly with large files so I'll just remove the file count.
If we really want a file count, we can just pass files in individually and count them on the front end.
I've taken out the file count and because the total number of files doesn't make much sense without it, I've taken that out as well.

Comment (MancunianSam) [reply src/index.ts]: Actually, the above explanation might not be totally accurate but there's definitely some async strangeness going on here that's preventing accurate reporting of which file we're up to for a particular percentage and as it's arguably not that valuable, I still think we should just take it out.

Comment (MancunianSam) [reply src/index.ts]: Ok, I think I was right although I'm not exactly sure why. I've removed the call to `files.map` and put in a good old fashioned for loop with another one for the chunks in the same function and this is now giving sensible results and it's probably easier to read as well.

The point still stands about whether we want to have a processed files count but now I've managed to get it working, I've changed my mind about it :smile:

File: .eslintrc.json (hunk 10 above)
Comment (suzannehamilton) [reply .eslintrc.json]: Cheers. 👍 

What does it mean to lint the code as ES6, when the code is written in TypeScript rather than JavaScript? Is there such a thing as ES5/ES6 TypeScript? Sorry if that's a very confused question. 🙂

File: src/index.ts (hunk 6 above)
Comment (suzannehamilton) [reply src/index.ts]: Cheers for looking into it! I haven't followed all the steps you've taken, but the current version looks like it gives consistent results so let's try it for now. 👍 We can come back to it if it looks like it might be more useful to have an overall percentage, or something else.

File: .eslintrc.json (hunk 10 above)
Comment (MancunianSam) [reply .eslintrc.json]: There's probably no such thing as ES5 Typescript but eslint will lint javascript as well as Typescript which is why the setting exists. 
As to why it's here, it's what came out of the defaults when I ran the setup script.
